src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs

```diff
@@ -97,21 +98,35 @@ public Task Invoke(HttpContext context)
                     case StaticFileContext.PreconditionState.ShouldProcess:
                         if (fileContext.IsHeadMethod)
                         {
-                            return fileContext.SendStatusAsync(Constants.Status200Ok);
+                            await fileContext.SendStatusAsync(Constants.Status200Ok);
+                            return;
                         }
                         if (fileContext.IsRangeRequest)
                         {
-                            return fileContext.SendRangeAsync();
+                            await fileContext.SendRangeAsync();
+                            return;
+                        }
+                        try
+                        {
+                            await fileContext.SendAsync();
+                            _logger.LogFileServed(fileContext.SubPath, fileContext.PhysicalPath);
+                            return;
+                        }
+                        catch (FileNotFoundException)
+                        {
+                            _logger.LogFileNotFound(fileContext.SubPath);
```

Comment:
This does not reset the headers set by SendAsync. Either reset the headers at this point or delay setting headers in SendAsync

Reply:
Good catch. I added test coverage for this and delayed setting headers until after the FileNotFoundException would have been thrown

Reply:
You didn't do this for the SendFileAsync code path, nor is it possible to delay setting the headers.